cpu/atmega_common/periph/adc.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2016 Laurent Navet <laurent.navet@gmail.com>
+ *               2017 HAW Hamburg, Dimitri Nahm
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_periph
+ * @{
+ *
+ * @file
+ * @brief       Low-level ADC driver implementation for ATmega family
+ *
+ * @author      Laurent Navet <laurent.navet@gmail.com>
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/adc.h"
+#include "periph_conf.h"
+
+#define ADC_MAX_CLK         (200000U)
+
+static mutex_t lock = MUTEX_INIT;
+
+static inline void prep(void)
+{
+    mutex_lock(&lock);
+    /* Enable ADC */
+    ADCSRA |= (1 << ADEN);
+}
+
+static inline void done(void)
+{
+    /* Disable ADC */
+    ADCSRA &= ~(1 << ADEN);
+    mutex_unlock(&lock);
+}
+
+int adc_init(adc_t line)
+{
+    uint32_t clk_div = 1;
+
+    /* check if the line is valid */
+    if (line >= ADC_NUMOF) {
+        return -1;
+    }
+
+    prep();
+
+    /* Disable corresponding Digital input */
+    #if defined (CPU_ATMEGA328P) || defined (CPU_ATMEGA1281)
+        DIDR0 |= (1 << line);
+    #endif
+    #ifdef CPU_ATMEGA2560
+        if(line < 8)
+            DIDR0 |= (1 << line);
+        else
+            DIDR2 |= (1 << (line-8));
+    #endif
+
+    /* Set ADC-pin as input */
+    #ifdef CPU_ATMEGA328P
+        DDRC &= ~(1 << line);
+        PORTC &= ~(1 << line);
+    #endif
+    #ifdef CPU_ATMEGA2560
+        if (line < 8) {
```

Comment:
maybe move upward to previous matching `if else`  simplifies preprocessor stuff a bit